### ISSUE TITLE
Feature Request: Jacob's `Delta-Bar-Delta` Update Rule For `GradientDescent` Optimizer

### ISSUE THREAD

**Issue by ranjodhsingh1729** [mlpack/ensmallen#437]
# Description
I was trying to implement t-SNE for mlpack and required the delta bar delta update rule given bellow.

<img width="774" height="455" alt="Image" src="https://github.com/user-attachments/assets/afaa7318-a419-495c-977c-ddd8a51a81db" />

I am currently using a callback to achieve the same result. It would be better if this was provided by ensmallen itself.

I can make a PR if there's interest for such a addition.

# References
[[1](https://www2.bcs.rochester.edu/sites/robbie/jacobs.nn88.pdf)] R.A. Jacobs. Increased rates of convergence through learning rate adaptation. Neural
Networks, 1:295–307, 1988.

**Comment by ranjodhsingh1729** [comment]
I wanted to refactor the gradient descent class to accept the update rule as a template parameter and add support for the delta-bar-delta update, which users could select instead of the default one.

I'm not entirely sure how well this works with SGD - I think it does, but I'll confirm. If it does, it would be great if both gradient descent and SGD could share the same update rule.

Initially, I used callbacks to modify the gradient before the optimizer applied the update - a hacky approach, but I was eager to see the algorithm working.

I ended up modifying the momentum update to implement this rule. It was really just a small change (4-5 lines), and I'm now using it with SGD (batch size = number of points).

t-SNE traditionally, and for good reason, uses the full batch gradient descent, so this change makes a lot of sense to me.

Callbacks are still handy for things like early exaggeration with reduced momentum, but I'm leaning toward handling that with two seperate optimizer instances with different paramters running one after the other since the hyperparameter adjustment only happens once.

I hope this clears things up, and apologies for not being clearer earlier.

**Comment by rcurtin** [comment]
No worries, there are a lot of design questions so there are tons of different ways we could go here, it all depended on what the needs were.  My recommendation would be that we add `DeltaBarDelta` as a new optimizer to ensmallen, and the implementation strategy could be like this:

 * Change `GradientDescent` to `GradientDescentType` and add `UpdateRuleType` as a template parameter.
 * Implement `VanillaUpdate` with the regular step type.
 * Implement `DeltaBarDeltaUpdate` with the new one.
 * Typedef `GradientDescent = GradientDescentType<VanillaUpdate>` and `DeltaBarDelta = GradientDescentType<DeltaBarDeltaUpdate>`
 * However, if `DeltaBarDeltaUpdate` has any parameters that the user needs to set (I am not sure, I have not checked closely), then we will need to do the same thing we do for SGD... where `DeltaBarDelta` is a class that internally holds a `GradientDescentType` and provides the same `Optimize()` method, but allows any parameters to be set in the constructor.  Take a look at `Adam`, for instance, to see what I mean.
 * Add documentation to `doc/optimizers.md`.  You can start with the `GradientDescent` documentation and adapt it accordingly.
 * Update the differentiable function optimizer list in `doc/function_types.md` to include `DeltaBarDelta`.
 * Add some tests for `DeltaBarDelta`: pretty simple, you can adapt the gradient descent tests.  They might need some tuning and since the delta-bar-delta rule is a bit smarter, you may be able to get them to converge in fewer iterations (hopefully).

I know that is a lot, but that's probably the best thing to do: even though our goal is just to use this in t-SNE in mlpack, if we can provide something more general that other ensmallen users can take advantage of, that's even better.  Let me know what you think.

**Comment by ranjodhsingh1729** [comment]
Hi, thanks for the detailed outline this made things really straightforward.

Just one question:
We can't implement another `VanillaUpdate` inside `ens` namespace. So for now i am just using the one in `SGD`. Do you think it would be better to keep separate `VanillaUpdate` implementations for `SGD` and `GradientDescentType`, under different names? 


> if we can provide something more general that other ensmallen users can take advantage of, that's even better.

I agree.

**Comment by ranjodhsingh1729** [comment]
Hi,

Is there a way for `GradientDescent` to report progress?

I noticed that `GradientDescent` doesn't work with `ProgressBar` as `ProgressBar` requires `batchSize` in the Optimizer and `NumFunctions` in the Function's Signature. So i tried using `PrintLoss` but it prints loss when `EndEpoch` is called which the `GradientDescent` Doesn't Call. Would it make sense for `GradientDescent` to call `EndEpoch` every iteration?
